File: java-springboot/src/main/java/com/william/travel/entity/Location.java
package com.william.travel.entity;

import jakarta.persistence.*;
import lombok.Getter;
import lombok.Setter;
import lombok.EqualsAndHashCode;
import lombok.ToString;

import java.math.BigDecimal;
import java.util.HashSet;
import java.util.Set;

@Entity
@Table(name = "locations")
@Getter
@Setter
@EqualsAndHashCode(callSuper = true, exclude = {"country", "region", "district", "ward", "category", "ratings", "media"})
@ToString(exclude = {"country", "region", "district", "ward", "category", "ratings", "media"})
public class Location extends BaseEntity {

    @Column(name = "name", length = 255, nullable = false)
    private String name;

    @Column(name = "name_code", length = 255)
    private String nameCode;

    @Column(name = "description", columnDefinition = "TEXT")
    private String description;

    @Column(name = "description_code", columnDefinition = "TEXT")
    private String descriptionCode;

    @Column(name = "latitude")
    private Double latitude;

    @Column(name = "longitude")
    private Double longitude;

    @Column(name = "geom", columnDefinition = "TEXT")
    private String geom;

    @Column(name = "address", length = 255)
    private String address;

    @Column(name = "city", length = 100)
    private String city;

    @Column(name = "thumbnail_url", length = 255)
    private String thumbnailUrl;

    @Column(name = "price_min", precision = 10, scale = 2)
    private BigDecimal priceMin;

    @Column(name = "price_max", precision = 10, scale = 2)
    private BigDecimal priceMax;

    @Column(name = "popularity_score")
    private Double popularityScore = 0.0;

    @Column(name = "is_active", nullable = false)
    private Boolean isActive = true;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "country_id")
    private Country country;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "region_id")
    private Region region;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "district_id")
    private District district;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "ward_id")
    private Ward ward;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "category_id", nullable = false)
    private LocationCategory category;

    @OneToMany(mappedBy = "location", fetch = FetchType.LAZY)
    private Set<Rating> ratings = new HashSet<>();

    @OneToMany(mappedBy = "location", fetch = FetchType.LAZY)
    private Set<Media> media = new HashSet<>();
}

File: be/app/schemas/location.py
from datetime import date
from typing import List, Optional

from pydantic import BaseModel, Field

from app.schemas.base import BaseSchema, DateMixin, TimestampMixin


class ContinentBase(BaseModel):
    name: str
    code: str
    name_code: Optional[str] = None
    background_image: Optional[str] = None
    logo: Optional[str] = None
    description: Optional[str] = None
    description_code: Optional[str] = None
    status: int = 1


class ContinentCreate(ContinentBase):
    pass


class ContinentUpdate(BaseModel):
    name: Optional[str] = None
    code: Optional[str] = None
    name_code: Optional[str] = None
    background_image: Optional[str] = None
    logo: Optional[str] = None
    description: Optional[str] = None
    description_code: Optional[str] = None
    status: Optional[int] = None


class Continent(ContinentBase, DateMixin, BaseSchema):
    id: int


class ContinentWithCountries(Continent):
    countries: List["Country"] = []


class CountryBase(BaseModel):
    name: str
    code: str
    name_code: Optional[str] = None
    description: Optional[str] = None
    description_code: Optional[str] = None
    background_image: Optional[str] = None
    logo: Optional[str] = None
    status: int = 1
    continent_id: int


class CountryCreate(CountryBase):
    pass


class CountryUpdate(BaseModel):
    name: Optional[str] = None
    code: Optional[str] = None
    name_code: Optional[str] = None
    description: Optional[str] = None
    description_code: Optional[str] = None
    background_image: Optional[str] = None
    logo: Optional[str] = None
    status: Optional[int] = None
    continent_id: Optional[int] = None


class Country(CountryBase, DateMixin, BaseSchema):
    id: int


class RegionBase(BaseModel):
    name: str
    code: str
    name_code: Optional[str] = None
    description: Optional[str] = None
    description_code: Optional[str] = None
    background_image: Optional[str] = None
    logo: Optional[str] = None
    status: int = 1
    country_id: int


class RegionCreate(RegionBase):
    pass


class RegionUpdate(BaseModel):
    name: Optional[str] = None
    code: Optional[str] = None
    name_code: Optional[str] = None
    description: Optional[str] = None
    description_code: Optional[str] = None
    background_image: Optional[str] = None
    logo: Optional[str] = None
    status: Optional[int] = None
    country_id: Optional[int] = None


class Region(RegionBase, DateMixin, BaseSchema):
    id: int


class DistrictBase(BaseModel):
    name: str
    code: str
    name_code: Optional[str] = None
    description: Optional[str] = None
    description_code: Optional[str] = None
    background_image: Optional[str] = None
    logo: Optional[str] = None
    status: int = 1
    region_id: int


class DistrictCreate(DistrictBase):
    pass


class DistrictUpdate(BaseModel):
    name: Optional[str] = None
    code: Optional[str] = None
    name_code: Optional[str] = None
    description: Optional[str] = None
    description_code: Optional[str] = None
    background_image: Optional[str] = None
    logo: Optional[str] = None
    status: Optional[int] = None
    region_id: Optional[int] = None


class District(DistrictBase, DateMixin, BaseSchema):
    id: int


class WardBase(BaseModel):
    name: str
    code: str
    name_code: Optional[str] = None
    description: Optional[str] = None
    description_code: Optional[str] = None
    background_image: Optional[str] = None
    logo: Optional[str] = None
    status: int = 1
    district_id: int


class WardCreate(WardBase):
    pass


class WardUpdate(BaseModel):
    name: Optional[str] = None
    code: Optional[str] = None
    name_code: Optional[str] = None
    description: Optional[str] = None
    description_code: Optional[str] = None
    background_image: Optional[str] = None
    logo: Optional[str] = None
    status: Optional[int] = None
    district_id: Optional[int] = None


class Ward(WardBase, DateMixin, BaseSchema):
    id: int


class LocationCategoryBase(BaseModel):
    name: str
    status: bool = True


class LocationCategoryCreate(LocationCategoryBase):
    pass


class LocationCategoryUpdate(BaseModel):
    name: Optional[str] = None
    status: Optional[bool] = None


class LocationCategory(LocationCategoryBase, TimestampMixin, BaseSchema):
    id: int


class LocationBase(BaseModel):
    name: str
    name_code: Optional[str] = None
    description: Optional[str] = None
    description_code: Optional[str] = None
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    geom: Optional[str] = None
    address: Optional[str] = None
    city: Optional[str] = None
    country_id: Optional[int] = None
    region_id: Optional[int] = None
    district_id: Optional[int] = None
    ward_id: Optional[int] = None
    category_id: int
    thumbnail_url: Optional[str] = None
    price_min: Optional[float] = None
    price_max: Optional[float] = None
    is_active: bool = True


class LocationCreate(LocationBase):
    pass


class LocationUpdate(BaseModel):
    name: Optional[str] = None
    name_code: Optional[str] = None
    description: Optional[str] = None
    description_code: Optional[str] = None
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    geom: Optional[str] = None
    address: Optional[str] = None
    city: Optional[str] = None
    country_id: Optional[int] = None
    region_id: Optional[int] = None
    district_id: Optional[int] = None
    ward_id: Optional[int] = None
    category_id: Optional[int] = None
    thumbnail_url: Optional[str] = None
    price_min: Optional[float] = None
    price_max: Optional[float] = None
    popularity_score: Optional[float] = None
    is_active: Optional[bool] = None


class Location(LocationBase, TimestampMixin, BaseSchema):
    id: int
    popularity_score: float = 0
    country: Optional[Country] = None
    region: Optional[Region] = None
    district: Optional[District] = None
    ward: Optional[Ward] = None
    category: Optional[LocationCategory] = None
